views: Support 'j/k', 'J/K' and 'G/end' for navigation in popups.

preetmishra opened the pull request:
This PR is a follow-up to #518. This will generalize navigation and make the user interface consistent.
@neiljp

preetmishra commented:
@neiljp Thanks for the review. I agree that there's duplication but there are two seperate views (Stream and Message). One possibility I can think of is making a base test class for popups, would that be fine?

neiljp commented:
@preetmishra Right, it would seem straightforward to do that, maybe an `InfoView`? (based on the current class names)

I would keep this code around, but this other approach may be simpler.

preetmishra commented:
Yes, that seems plausible. I'll start working on it.

preetmishra commented:
@neiljp I have added an `InfoView` which encapsulates the basic tests for any popup and leveraged it to simplify `TestStream/MsgInfoView` tests. 

I have also added another commit to refactor `TestHelpMenu` to leverage `InfoView`.

neiljp commented:
@preetmishra Sorry if I wasn't clear; I was meaning refactoring the **code** to use another class with the common features, not the tests :)

preetmishra commented:
@neiljp Thanks. I have added `InfoView` with the common method and inherited it in the other `Stream/MsgInfoView`.

The second commit is a suggested refactor of `HelpView` to leverage `InfoView` class.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -751,6 +751,24 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
+    def keypress(self, size: urwid_Size, key: str) -> str:
+        if is_command_key('GO_BACK', key) \
+                or is_command_key(self.command, key):
+            self.controller.exit_popup()
```
`self.controller` and `self.command` are just assumed to be present here; I think it would be clearer if we just had an initializer which specified the parameters it needed and were used at each call point.

preetmishra commented:
@neiljp Thanks for the review. The sequence you suggested is indeed much cleaner than what I had proposed previously. I have pushed the new changes. 

Also, I renamed `InfoView` to `PopUpView` as it seemed more consistent with HelpView and Stream/MsgInfoView (`InfoView` seemed only limited to infoviews).

preetmishra commented:
@neiljp I have added another refactor commit to introduce a fixture which will help in reducing code duplication. I think it is apt to introduce it here before the 'new' feature since they are going to use it in `keypress_navigation` tests. However, the commit heading already got too long with 'tests', 'areas' and 'refactor' that I had to settle with a short heading that is not as descriptive.

For the rest of the two commits, I have made the changes that were suggested. I have also renamed `help_menu_content` to `log` in `HelpView` to make it consistent with `Msg/StreamInfoView'.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -754,28 +754,49 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
 
-class HelpView(urwid.ListBox):
-    def __init__(self, controller: Any) -> None:
+class PopUpView(urwid.ListBox):
```
This class should be simple to add a test for :)

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -895,18 +887,12 @@ def __init__(self, controller: Any, msg: Message) -> None:
 
-        self.log = urwid.SimpleListWalker(
-            [urwid.AttrWrap(
+        log = [urwid.AttrWrap(
```
Since this is no longer a `...Walker`, let's name it more descriptively, as per my other comment.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -754,28 +754,49 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
         self.width = sum(max_widths)
 
-        help_menu_content = []  # type: List[urwid.AttrWrap]
+        log = []  # type: List[urwid.AttrWrap]
```
Thanks. I'll keep that in mind.

preetmishra commented:
@neiljp Thanks. I have added a test for `PopUpView`'s init method as its `keypress` is being tested in the derived classes anyway. I have also renamed the variables as per your suggestions.

neiljp commented:
I understand the keypress is being tested in the derived classes, but the derived classes may change behavior later; I'm not sure this is much overhead or a challenge? :)

preetmishra commented:
@neiljp My bad, I didn't consider the fact that the derived classes may change the behavior later. Thanks. I have added the tests for different keypress options as well.

Other than that, I have added another refactor commit to change `super_view` to `super_keypress` 
in `HelpView`'s test as the latter better defines what the variable holds.

Also, the commit order that GitHub is showing isn't what I had locally. The order that I pushed was 4 -> 3 -> 2 -> 1.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1093,6 +1093,53 @@ def test_topics_view(self, mocker, stream_button, width=40):
+    def pop_up_view(self, mocker):
+        self.controller = mocker.Mock()
+        self.command = ''
+        self.widgets = []
```
Probably these should have some kind of meaningful values.

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1093,6 +1093,53 @@ def test_topics_view(self, mocker, stream_button, width=40):
+        self.pop_up_view.keypress((200, 20), 'cmd_key')
+        assert self.controller.exit_popup.called
+
+    def test_keypress_navigation_k(self, mocker,
```
My bad, I had to test this function specifically so I appended '_k'. I'll fix this.

neiljp replied:
For reference, you can try the `-k` parameter to `pytest` with something like `PopUp and navig` to resolve this :)

preetmishra replied:
Thanks, this would be really helpful in the future.

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1093,6 +1093,53 @@ def test_topics_view(self, mocker, stream_button, width=40):
+    def test_init(self):
+        assert self.pop_up_view.controller == self.controller
+        assert self.pop_up_view.command == self.command
+        self.list_walker.assert_called_once_with(self.widgets)
```
The other similar `__init__` tests don't have super-init asserting and the previous mocked super-init method was raising issues for the `keypress` method so I removed it. Do we need it?

neiljp replied:
Yes, let's leave this for now.

preetmishra replied:
Alright. I'll add it back.

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1093,6 +1093,53 @@ def test_topics_view(self, mocker, stream_button, width=40):
+        self.pop_up_view.keypress((200, 20), key)
+        assert self.controller.exit_popup.called
+
+    def test_keypress_command_key(self, mocker):
```
The `GO_BACK` test has parameterization and doesn't require any mocking. I think it is better to keep it separated?

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1093,6 +1093,53 @@ def test_topics_view(self, mocker, stream_button, width=40):
+            lambda command, key:
+            False if command == self.command
+            else is_command_key(command, key))
+        )
```
The `PopUpView`'s `keypress` has a condition check - `is_command_key(self.command, key)`, if left without a patch, this raises an `Invalid Command` exception as there is no command for `self.command=''`. I patched it to deal with that case.

neiljp replied:
Ah, yes - I'm not saying it doesn't need a patch, but that this test is regarding what should happen if the command matches whatever is passed in. We have a test which guarantees this already, so we don't need a variable return value?

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -1093,6 +1093,53 @@ def test_topics_view(self, mocker, stream_button, width=40):
+        mocker.patch(VIEWS + '.is_command_key', side_effect=(
+            lambda command, key:
+            key in KEY_BINDING[self.command] if command == self.command
+            else False)
```
In order to check if the `exit_popup` gets triggered specifically by the key in `self.command` set, I wrote this patch. Where exactly are you suggesting to use `self.command`?

neiljp replied:
I understand why you're mocking this now, but can we not just use `self.command`? I see you're using a data-structure like in `keys.py`, but then just accessing it locally, and it's truthy, so does this just reduce to the command sub-expression?

neiljp replied:
(the key doesn't matter - we're not testing behavior of keys here, but the command, which could map to various keys)

preetmishra replied:
I understand what you're suggesting, I'll make the amendments.

preetmishra commented:
@neiljp Thanks. I'll revert the `size` variable back in the tests. I also did this in #539 as per your suggestion to use default parameter as a compromise for `size` variable, I'll make these changes there as well.

preetmishra commented:
@neiljp Thanks. I have made the amendments in the test as per your suggestions. I also rewrote the commit message to include the `Walker` usage change. It is ready for review.

neiljp commented:
@preetmishra Thanks for this! I just merged it after some very minor tweaks :tada:

There were no code changes, I just moved one import from one commit to another, as you'd introduced an import in an earlier commit than it was required.

preetmishra commented:
Thanks so much! :tada:
